packages/protocol/contracts/L1/libs/LibProposing.sol

```diff
@@ -289,4 +231,87 @@ library LibProposing {
             hash := keccak256(inputs, mul(5, 32))
         }
     }
+
+    function hashAssignmentForTxList(
+        TaikoData.ProverAssignment memory assignment,
+        bytes32 txListHash
+    )
+        internal
+        pure
+        returns (bytes32)
+    {
+        return keccak256(
+            abi.encode(
+                "PROVER_ASSIGNMENT",
+                txListHash,
+                assignment.feeToken,
+                assignment.expiry,
+                assignment.tierFees
+            )
+        );
+    }
+
+    function _validateAssignment(
+        uint16 minTier,
+        bytes32 txListHash,
+        TaikoData.ProverAssignment memory assignment
+    )
+        private
+        returns (uint256 proverFee)
+    {
+        // Check assignment not expired
+        if (block.timestamp >= assignment.expiry) {
+            revert L1_ASSIGNMENT_EXPIRED();
+        }
+
+        if (txListHash == 0 || assignment.prover == address(0)) {
+            revert L1_ASSIGNMENT_INVALID_PARAMS();
+        }
+
+        // Hash the assignment with the txListHash, this hash will be signed by
+        // the prover, therefore, we add a string as a prefix.
+        bytes32 hash = hashAssignmentForTxList(assignment, txListHash);
+
+        if (!assignment.prover.isValidSignature(hash, assignment.signature)) {
+            revert L1_ASSIGNMENT_INVALID_SIG();
+        }
+
+        // Find the prover fee using the minimal tier
+        proverFee = _getProverFee(assignment.tierFees, minTier);
+
+        // The proposer irrevocably pays a fee to the assigned prover, either in
+        // Ether or ERC20 tokens.
+        if (assignment.feeToken == address(0)) {
```

Comment:
So right now, we also support ERC20 for 'single, EOA provers' not contracts-only ? I like this design.

Reply:
Yes.